### PR TITLE
Create movie details page and custom twig components

### PR DESCRIPTION
**Pull request by damyancodixis**
### What
Create movie details page and custom twig components

### How
- Create new page that displays details of a single movie
- Add custom query to `ReviewRepository` that fetches all reviews(not ratings) for a given movie 
- Improve flexibility on pagination component and utilize it for reviews section
- Create a json field rendering component
- Create a 5-star rating component
- Make custom twig extension to display large numbers in a short format

### Screenshots
Screenshot 2023-11-07 at 16 23 41
Screenshot 2023-11-07 at 16 23 56
Screenshot 2023-11-07 at 16 28 07

### CODE REVIEW

File: templates/components/pagination.html.twig [hunk 1]
```diff
@@ -1,80 +1,77 @@
...
+    perPage - number of objects per page
+    totalItems - the total number of objects in the current table
+    totalPages - the total number of pages for the table
+    routeName - the name propery of the #Route attribute of the controller
+    urlParams - all parameters that come after the base route url ('id' for a single movie page, etc.)
+#}
+<div class="flex items-center justify-between border-t border-gray-700 px-4 py-3 sm:px-6">
+    <div class="flex flex-1 justify-between sm:hidden">
+        <a href="{{ path(routeName, app.request.query.all|merge(urlParams, {'page': page - 1})) }}" class="relative inline-flex items-center rounded-md border border-gray-700 px-4 py-2 text-sm font-medium text-gray-700 hover:bg-gray-50 {{ page < 2 ? 'pointer-events-none bg-gray-500' : 'bg-white' }} ">Previous</a>
+        <a href="{{ path(routeName, app.request.query.all|merge(urlParams, {'page': page + 1})) }}" class="relative ml-3 inline-flex items-center rounded-md border border-gray-700 px-4 py-2 text-sm font-medium text-gray-700 hover:bg-gray-50 {{ page >= totalPages ? 'pointer-events-none bg-gray-500' : 'bg-white' }}">Next</a>
```
Comment (IrinaCodixis) [inline templates/components/pagination.html.twig]: Instead of passing everything in a controller, add params in the Include

File: src/Entity/Review.php [hunk 2]
```diff
@@ -116,9 +116,14 @@ public function getCreatedAt(): ?\DateTimeImmutable
         return $this->createdAt;
     }
 
-    public function setCreatedAt(\DateTimeImmutable $createdAt): static
+    public function setCreatedAt(\DateTimeImmutable|\DateTime $createdAt): static
     {
-        $this->createdAt = $createdAt;
+        if ($createdAt instanceof \DateTime) {
```
Comment (IrinaCodixis) [inline src/Entity/Review.php:121]: Why did you add this logic?

File: src/Repository/ReviewRepository.php [hunk 3]
```diff
@@ -12,4 +13,20 @@ public function __construct(ManagerRegistry $registry)
...
     }
+
+    public function findReviewsByMovie(string $movieId, int $page, int $perPage): Paginator
+    {
+        $queryBuilder = $this
+            ->createQueryBuilder('r')
+            ->andWhere('r.movie = :movieId')
+            ->setParameter('movieId', $movieId)
+            ->andWhere('r.title IS NOT NULL')
+            ->setFirstResult(($page - 1) * $perPage)
```
Comment (IrinaCodixis) [inline src/Repository/ReviewRepository.php:25]: You don't order them by date?

File: src/Twig/AppExtension.php [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+    public function shortenNumber(int $number): string
+    {
+        if ($number >= 1_000_000) {
+            return round($number / 1_000_000, 1) . 'M';
+        }
+        else if ($number >= 100_000) {
+            return round($number / 1000) . 'K';
+        }
+        else {
```
Comment (IrinaCodixis) [inline src/Twig/AppExtension.php]: Generally, in this situation, linter would complain and say that else is unnecessary, you can put last return after your if statements.
Plus, I think else if is unnecessary if you are returning, it can be just 2 if statements

File: templates/components/rate-menu.html.twig [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        currentLabel() {
+            let r = this.rating;
+            if (this.hoverRating != this.rating) {
+                r = this.hoverRating;
+            }  
+            let i = this.ratings.findIndex(e => e.amount == r);
+            if (i >=0) {
+                return this.ratings[i].label;
+            } 
+            else {
```
Comment (IrinaCodixis) [inline templates/components/rate-menu.html.twig]: Same here - unnecessary if

File: templates/pages/movie.html.twig [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+                            <button type="button" class="inline-flex w-full justify-center rounded-md bg-indigo-600 px-6 py-3 text-lg font-semibold text-white shadow-sm hover:bg-indigo-500 focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-indigo-600">
+                                Rate
+                            </button>
+                        </div>
+                    </div>
+                </div>  
+            </div>
+        </div>
+
+        {# Review menu #}
```
Comment (IrinaCodixis) [inline templates/pages/movie.html.twig]: I am confused with working 'menu'. Is it a dialog? Then it should be a dialog or modal

File: templates/pages/movie.html.twig [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+                        
+                            <div>
+                                <label for="comment" class="block text-sm font-medium leading-6">Review</label>
+                                <div class="mt-2">
+                                    <textarea rows="4" name="comment" id="comment" class="bg-gray-800 block w-full rounded-md border-0 py-1.5 text-gray-300 shadow-sm ring-1 ring-inset  placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-indigo-600 sm:text-sm sm:leading-6"></textarea>
+                                </div>
+                            </div>
+
+                            <div name="rating" class="mt-2">
+                                {% include "components/rate-menu.html.twig" %}
```
Comment (IrinaCodixis) [inline templates/pages/movie.html.twig]: Here menu as well?

**Review by IrinaCodixis — CHANGES_REQUESTED** [review]

File: src/Entity/Review.php (hunk 2 above)
Comment (damyancodixis) [reply src/Entity/Review.php:121]: I added it because if type `DateTime` was passed to the `createdAt` property, it complained since it's not of type `DateTimeImmutable`. Because of this I wasn't able to use the `<dateTimeBetween()>` faker function in `fixtures/review.yaml`

File: templates/pages/movie.html.twig (hunk 7 above)
Comment (damyancodixis) [reply templates/pages/movie.html.twig]: I thought having the option to give a rating when a submitting a review would be suitable.

Comment (IrinaCodixis) [reply templates/pages/movie.html.twig]: I am talking more about the naming of the component. These rating stars don't seem like a menu. The wording is confusing

File: src/Entity/Review.php (hunk 2 above)
Comment (IrinaCodixis) [reply src/Entity/Review.php:121]: Just to test, could you revert this change and try:
'<(\DateTimeImmutable::createFromMutable(dateTimeBetween("-20 years", "now")))>'

Comment (damyancodixis) [reply src/Entity/Review.php:121]: I already tried using the `createFromMutable()` function, but for some reason it still complained that the type received was `DateTime`. Reverting the changes and copy-pasting the line you suggested triggered the same error:
Screenshot 2023-11-08 at 16 00 28

**Review by IrinaCodixis — APPROVED** [review]
